lib/rom/rails/railtie.rb

```diff
@@ -11,42 +11,36 @@
 module ROM
   module Rails
     class Railtie < ::Rails::Railtie
-      def self.disconnect
-        return unless ROM.env
-
-        ROM.env.repositories.each_value do |repository|
-          repository.adapter.disconnect
+      initializer 'rom.configure_action_controller' do
+        ActiveSupport.on_load(:action_controller) do
+          ActionController::Base.send(:include, ControllerExtension)
         end
       end
 
-      def self.load_all
-        load_schema
-
-        %w(relations mappers commands).each do |type|
-          load_files(type, ::Rails.root)
-        end
+      # Make `ROM::Rails::Configuration` instance available to the user via
+      # `Rails.application.config` before other initializers run.
+      config.before_initialize do |_app|
+        config.rom = Configuration.new
       end
 
-      def self.load_files(type, root)
-        Dir[root.join("app/#{type}/**/*.rb").to_s].each do |path|
-          load(path)
-        end
+      # Reload ROM-related application code on each request.
+      config.to_prepare do |_config|
+        Railtie.reload if ROM.env
       end
 
-      def load_schema
-        load(schema_file) if schema_file.exist?
-      end
+      # This is where the initial setup of ROM occurs.
+      # TODO: Freeze the configuration.
+      config.after_initialize do |app|
+        # At this point ActiveRecord::Base.configurations are already populated,
+        # but ROM should NOT be yet. Will only be called once.
+        fail if ROM.env
```

Comment:
What exception will the user see here? Should it be more descriptive? Some hint as to what has happened.